File: 43 Multiply Strings/43 LL.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

/*
https://leetcode.com/problems/multiply-strings/

Given two numbers represented as strings, return multiplication of the numbers as a string.

Note: The numbers can be arbitrarily large and are non-negative.
*/

class Solution
{
private:
	void selfadding(string &rvanswer, string &rvb)
	{
		int carry = 0;
		string::iterator data = rvanswer.begin();
		string::iterator adder = rvb.begin();
		for (; data != rvanswer.end(); ++data)
		{
			carry += int(data[0] - '0');
			if (adder != rvb.end())
			{
				carry += int(adder[0] - '0');
				++adder;
			}
			if (carry >= 10)
			{
				data[0] = char('0' + carry - 10);
				carry = 1;
			}
			else
			{
				data[0] = char('0' + carry);
				carry = 0;
			}
		}
		for (; adder != rvb.end(); ++adder)
		{
			carry += int(adder[0] - '0');
			if (carry >= 10)
			{
				rvanswer += char('0' + carry - 10);
				carry = 1;
			}
			else
			{
				rvanswer += char('0' + carry);
				carry = 0;
			}
		}
		if (carry != 0) rvanswer += char('0' + carry);
	}
public:
	string multiply(string num1, string num2)
	{
		string result = "0";
		if ((num1 == "") || (num1 == "0") || (num2 == "") || (num2 == "0"))
		{
			return result;
		}
		if (num1.size() < num2.size())
		{
			num1.swap(num2);
		}
		reverse(num1.begin(), num1.end());
		reverse(num2.begin(), num2.end());
		vector<string> multable(9, num1);
		for (int i = 1; i < multable.size(); ++i)
		{
			selfadding(multable[i], multable[i - 1]);
		}
		string zeros;
		for (int i = 0; i < num2.length(); i++)
		{
			if (num2[i] != '0')
			{
				zeros += multable[int(num2[i] - '1')];
				selfadding(result, zeros);
			}
			zeros.resize(i + 1);
			zeros[i] = '0';
		}
		reverse(result.begin(), result.end());
		return result;
	}
};

/*
idea: simulate hand multiply
complexity: O(N)
*/

int main(void)
{
	Solution engine;
	cout << engine.multiply("123", "999999999999999999999") << '\n';
	return 0;
}

File: 43 Multiply Strings/43 LL.py
class Solution(object):
	def multiply(self, num1, num2):
		num1 = [ord(x) - ord('0') for x in reversed(num1)]
		num2 = [ord(x) - ord('0') for x in reversed(num2)]
		if len(num1) == 0 or (len(num1) == 1 and num1[0] == 0):
			return str('0')
		if len(num2) == 0 or (len(num2) == 1 and num2[0] == 0):
			return str('0')
		result = [0] * (len(num1) + len(num2))
		for i in xrange(len(num1)):
			for j in xrange(len(num2)):
				result[i + j] += num1[i] * num2[j]
		pos = 0
		carry = 0
		while pos < len(result):
			carry += result[pos]
			result[pos] = chr(carry % 10 + ord('0'))
			carry /= 10
			pos += 1
		while carry != 0:
			result.append(chr(carry % 10 + ord('0')))
			carry /= 10
		while result[-1] == '0':
			result.pop()
		return ''.join(reversed(result))

engine = Solution()
print engine.multiply("12", "9999999999")